File: d2/d2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readLines() []string {
	wd, _ := os.Getwd()
	read := wd + "\\input.txt"
	file, _ := os.Open(read)
	scanner := bufio.NewScanner(file)
	lines := make([]string, 0)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func main() {
	lines := readLines()
	m := make(map[string]int)
	for i := 0; i < len(lines); i++ {
		split := strings.Fields(lines[i])
		cmd := split[0]
		val, _ := strconv.Atoi(split[1])
		switch cmd {
		case "forward":
			m["pos"] += val
			m["depth"] += m["aim"] * val
		case "up":
			m["aim"] -= val
		case "down":
			m["aim"] += val
		}
	}
	fmt.Printf("Solution part 1 %d, part 2 %d", m["aim"]*m["pos"], m["depth"]*m["pos"])
}

File: solutions/d2/d2.py
instructions = open("C:/Users/simon/Documents/aoc2021/d2/input.txt").readlines()

def get_state(part_2=False):
    state = {'aim': 0, 'pos': 0, 'depth': 0}
    for instr in instructions:
        dir, val = instr.split(' ')
        val = int(val)
        if dir == 'up':
            state['aim'] -= val
        elif dir == 'forward':
            state['pos'] += val
            state['depth'] += val*state['aim']
        elif dir == 'down':
            state['aim'] += val
    ret_val = state['aim']*state['pos'] if not part_2 else state['depth']*state['pos']
    return ret_val

print(f"Solution part 1: {get_state()}, part 2: {get_state(True)}")